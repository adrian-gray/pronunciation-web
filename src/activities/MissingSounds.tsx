import React, { useState } from "react";

import MemberGate from "./../components/MemberGate";
import SplitHilite from "./../components/SplitHilite";
import Tile from "./../components/Tile";

interface IMissingSoundsProps {
  isUserAuth: boolean;
  tag: string;
  ipa: string;
  words: string[];
  answers: string[];
}

export default (props: IMissingSoundsProps) => {
  const [isCorrect, setIsCorrect] = useState(
    Array(props.words.length).fill(undefined)
  );

  const { ipa, tag, words, answers } = props;
  const title = (
    <SplitHilite
      str={`Select the words that are ~MISSING~ the ${tag} – ${ipa} sound.`}
    />
  );

  let hasIncorrect = false;

  function checkAllCorrect(correct: boolean[]) {
    let numRemaining = answers.length;

    for (let i = 0; i < words.length; i++) {
      if (correct[i] === true) {
        numRemaining--;
      } else if (correct[i] === false) {
        hasIncorrect = true;
        return false;
      }
    }
    return numRemaining === 0;
  }

  const handleClick = (index: number) => {
    if (!props.isUserAuth) {
      return;
    }
    const isAnswerCorrect = answers.includes(words[index]);
    const newStateIsCorrect = isCorrect.slice();

    if (isCorrect[index] !== undefined) {
      newStateIsCorrect[index] = undefined;
    } else {
      newStateIsCorrect[index] = isAnswerCorrect;
    }

    checkAllCorrect(newStateIsCorrect);
    setIsCorrect(newStateIsCorrect);
  };

  const wordTiles = words.map((word: string, index: number) => (
    <Tile
      word={word}
      key={index}
      index={index}
      isCorrect={isCorrect[index]}
      handleClick={handleClick}
    />
  ));

  let result;
  checkAllCorrect(isCorrect);
  const numCorrect = isCorrect.filter(el => el).length;
  const areAllCorrectSelected = numCorrect === answers.length;

  let classBG = null;
  if (areAllCorrectSelected) {
    if (hasIncorrect) {
      classBG = "incorrect-bg";
      result = <h3 className="text-center">Try again</h3>;
    } else {
      classBG = "correct-bg";
      result = <h3 className="text-center">Yes ✓</h3>;
    }
  }

  const display = (
    <div className={`sub-canvas padded ${classBG}`}>
      {wordTiles}
      <div className="clear-float" />
      {result}
    </div>
  );

  return (
    <div className="headspace">
      {title}
      <MemberGate content={display} isUserAuth={props.isUserAuth} />
    </div>
  );
};
