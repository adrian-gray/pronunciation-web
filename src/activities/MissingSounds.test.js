import React from "react";
import { render, cleanup } from "@testing-library/react";

import MissingSounds from "./MissingSounds";

afterEach(cleanup);

test("MissingSounds", async () => {
  const props = {
    ipa: "a:",
    tag: "short-a",
    words: [
      "ran",
      "answer",
      "band",
      "start",
      "axe",
      "all",
      "thanks",
      "black",
      "table",
      "name"
    ],
    answers: ["start", "all", "table", "name"],
    isUserAuth: true
  };

  const { container } = render(<MissingSounds {...props} />);

  expect(container).toMatchSnapshot();
});
