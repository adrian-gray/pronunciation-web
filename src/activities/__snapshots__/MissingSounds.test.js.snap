// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissingSounds 1`] = `
<div>
  <div
    class="headspace"
  >
    <h3>
      <span>
        Select the words that are 
      </span>
      <span
        class="hilite"
      >
        MISSING
      </span>
      <span>
         the short-a â€“ a: sound.
      </span>
       
    </h3>
    <div
      class="sub-canvas padded null"
    >
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              ran
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              answer
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              band
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              start
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              axe
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              all
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              thanks
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              black
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              table
            </span>
          </p>
        </div>
      </div>
      <div
        class="cursor-pointer"
      >
        <div
          class="tile default-bg"
        >
          <p
            class="large-text"
          >
            <span>
              name
            </span>
          </p>
        </div>
      </div>
      <div
        class="clear-float"
      />
    </div>
  </div>
</div>
`;
